File: Design Patterns/Adapter Pattern/adapter.ts
// This code demonstrates the Adapter pattern:
// 1. We have an old printer interface (OldPrinter) and a new printer interface (NewPrinter).
// 2. We create an adapter (PrinterAdapter) that implements the new interface.
// 3. The adapter internally uses an instance of the old printer.
// 4. This allows us to use the old printer through the new interface.

// Old interface
interface OldPrinter {
    print(s: string): void;
}

// New interface
interface NewPrinter {
    printDocument(s: string): void;
}

// Old class
class OldPrinterImpl implements OldPrinter {
    print(s: string): void {
        console.log(`Old Printer: ${s}`);
    }
}

// New Adapter class
class PrinterAdapter implements NewPrinter {
    private oldPrinter: OldPrinter;

    constructor(oldPrinter: OldPrinter) {
        this.oldPrinter = oldPrinter;
    }

    printDocument(s: string): void {
        this.oldPrinter.print(s);
    }
}

// Client code
const oldPrinter = new OldPrinterImpl();
const newPrinter: NewPrinter = new PrinterAdapter(oldPrinter);

newPrinter.printDocument("Hello, Adapter!");

File: Design Patterns/Adapter Pattern/adapter.py
# This code demonstrates the Adapter pattern:
# 1. We have an old printer interface (OldPrinter) and a new printer interface (NewPrinter).
# 2. We create an adapter (PrinterAdapter) that implements the new interface.
# 3. The adapter internally uses an instance of the old printer.
# 4. This allows us to use the old printer through the new interface.

from abc import ABC, abstractmethod

# Existing interface
class OldPrinter(ABC):
    @abstractmethod
    def print(self, s: str) -> None:
        pass

# New interface
class NewPrinter(ABC):
    @abstractmethod
    def print_document(self, s: str) -> None:
        pass

# Existing class
class OldPrinterImpl(OldPrinter):
    def print(self, s: str) -> None:
        print(f"Old Printer: {s}")

# Adapter class
class PrinterAdapter(NewPrinter):
    def __init__(self, old_printer: OldPrinter):
        self.old_printer = old_printer

    def print_document(self, s: str) -> None:
        self.old_printer.print(s)

# Client code
if __name__ == "__main__":
    old_printer = OldPrinterImpl()
    new_printer = PrinterAdapter(old_printer)

    new_printer.print_document("Hello, Adapter!")